api/python/tests/integration/test_packages.py

```diff
@@ -623,6 +623,15 @@ def test_s3_set_dir(self):
 
             list_object_versions_mock.assert_called_with('bucket', 'foo/')
 
+    def test_set_dir_root_folder_named_slash(self):
+        with patch('quilt3.packages.list_object_versions') as list_object_versions_mock:
+            list_object_versions_mock.return_value = ([
+                dict(Key='/foo/a.txt', VersionId='xyz', IsLatest=True, Size=10),
+            ], [])
+            pkg = Package()
+            pkg.set_dir('bar', 's3://bucket//foo')  # top-level '/' folder
+            assert True, "No exception raised for top-level folder named '/'"
```

Comment:
that's not enough
at least you need to check physical key, because you could for example just strip leading slash and it won't raise exception but physical key would be wrong
I suggest check the whole entry like [here](https://github.com/quiltdata/quilt/blob/b0c1e92fb4426c4eab37e3f624dccd211a9cca13/api/python/tests/integration/test_packages.py#L620-L622)

Reply:
also you should call `list_object_versions_mock.assert_called_once_with()`

Reply:
Ouch, good catch. Looks like `urlunparse` automatically strips the 'invalid' leading '/'.
This is the simplest way I found to detect and revert it. Any better suggestions?
```
            result = urlunparse(('s3', self.bucket, quote(self.path), None, urlencode(params), None))
            missing_slash = f"s3://{self.bucket}{quote(self.path)}"
            if self.path and result.startswith(missing_slash):
                result = result.replace(missing_slash, f"s3://{self.bucket}/{quote(self.path)}")
            return result
```

Reply:
> Looks like urlunparse automatically strips the 'invalid' leading '/'.

it doesn't, it appears to add missing slash instead and we pass path without leading slash because we remove it earlier here
https://github.com/quiltdata/quilt/blob/b6c708faf6a3bc09c886bcddccf78198b3c85f45/api/python/quilt3/util.py#L156


so it looks like you should simply always add leading slash

Reply:
I _think_ line 156 is because "urlparse always returns absolute paths" , as otherwise `self.path` wouldn't have a leading slash in my case.

At any rate, if you can think of a better way to fix the error, feel free to improve the code and merge it.

Reply:
> I think line 156 is because "urlparse always returns absolute paths"

and we remove single leading slash so it become S3 key

> as otherwise self.path wouldn't have a leading slash in my case.

in your case there are *two* leading slashes in path, so after removing one you still have another

> if you can think of a better way to fix the error

I suggested one above

> feel free to improve the code and merge it.

okay